pkg/volume/util/fs/fs.go

```diff
@@ -55,28 +58,78 @@ func FsInfo(path string) (int64, int64, int64, int64, int64, int64, error) {
 	return available, capacity, usage, inodes, inodesFree, inodesUsed, nil
 }
 
+const (
+	// The block size in bytes.
+	statBlockSize uint64 = 512
+)
+
 // DiskUsage gets disk usage of specified path.
-func DiskUsage(path string) (*resource.Quantity, error) {
-	// First check whether the quota system knows about this directory
-	// A nil quantity with no error means that the path does not support quotas
-	// and we should use other mechanisms.
-	data, err := fsquota.GetConsumption(path)
-	if data != nil {
-		return data, nil
-	} else if err != nil {
-		return nil, fmt.Errorf("unable to retrieve disk consumption via quota for %s: %v", path, err)
+func DiskUsage(dir string) (*resource.Quantity, error) {
```

Comment:
Rather than copy logic from cAdvisor, please just call the function directly: https://github.com/google/cadvisor/blob/master/fs/fs.go#L551.  We would have to combine the DiskUsage and Find functions.

Reply:
I tried the following change:
```
diff --git a/pkg/volume/util/fs/BUILD b/pkg/volume/util/fs/BUILD
index 07da6389c2b..0d967c2aacc 100644
--- a/pkg/volume/util/fs/BUILD
+++ b/pkg/volume/util/fs/BUILD
@@ -55,7 +55,9 @@ go_library(
             "//staging/src/k8s.io/apimachinery/pkg/api/resource:go_default_library",
             "//vendor/golang.org/x/sys/windows:go_default_library",
         ],
-        "//conditions:default": [],
+        "//conditions:default": [
+        "//vendor/github.com/google/cadvisor/fs:go_default_library",
+        ],
     }),
 )

diff --git a/pkg/volume/util/fs/fs.go b/pkg/volume/util/fs/fs.go
index 2796e93d19e..510d6c34629 100644
--- a/pkg/volume/util/fs/fs.go
+++ b/pkg/volume/util/fs/fs.go
@@ -25,6 +25,7 @@ import (
        "strings"

        "golang.org/x/sys/unix"
+       cadvisorfs "github.com/google/cadvisor/fs"

        "k8s.io/apimachinery/pkg/api/resource"
        "k8s.io/kubernetes/pkg/volume/util/fsquota"
```
However, inside DiskUsage(), I am still looking for way to reference GetDirUsage.
cadvisorfs.GetDirUsage is reported as unresolved.

Reply:
What is the error?

Reply:
The error was that cadvisorfs.GetDirUsage is unresolved.

Looks like I need to access the func via:
```
type FsInfo interface {
   // GetDirUsage returns a usage information for 'dir'.
   GetDirUsage(dir string) (UsageInfo, error)
```
However, IDE tells me that there is not enough argument to GetDirUsage():
```
rootUsage, rootErr := cadvisorfs.FsInfo.GetDirUsage(path)
```

Reply:
vendor/github.com/google/cadvisor/fs/fs.go has this:
```
func NewFsInfo(context Context) (FsInfo, error) {
```
But cadvisorfs.NewFsInfo() gives unresolved compilation error as well.

Reply:
Here is the error when I add the call to cadvisorfs.NewFsInfo():
```
# k8s.io/kubernetes/pkg/volume/util/fs
./fs.go:62:2: undefined: fs.NewFsInfo
```
I think the reason maybe that DiskUsage() is inside fs package.
go compiler tries to resolve NewFsInfo() inside fs package (not of cadvisor). This happens even with import alias.

Reply:
@dashpole 
Please see my comments above.
Let me know if my understanding is correct.

Reply:
You shouldn't need to use the FsInfo interface... There is a stand-alone version as well as one associated with an FsInfo.  I was able to get it to build: https://github.com/kubernetes/kubernetes/compare/master...dashpole:no_du

The only problem I can see is that the cross-build will fail.  fs.go in cAdvisor is linux only, whereas fs.go in the volume util package is linux and darwin.  We either need to update cAdvisor to allow building that file with darwin (and confirm somehow that it actually works on darwin), or split pkg/volume/util/fs/fs.go into two files, one for darwin, and one for linux.  But splitting the file doesn't solve our security issue...

Reply:
Thanks for the hint.
```
+       resource
+       return &used, nil
```
There is some more work to be done for pkg/volume/util/fs/fs_windows.go

Reply:
ah, yeah.  I didn't do a cross build since I knew it would fail...